From 208a76bea4f80d4158e9a47480a8d7c564446984 Mon Sep 17 00:00:00 2001
From: Bryan Evenson <bevenson@melinkcorp.com>
Date: Tue, 10 Jun 2014 13:08:18 -0400
Subject: [PATCH 1/1] Adjusted makefiles for cross-compile.

Certain build variables, such as CFLAGS, are already defined by
Bitbake and should not be overridden.  This patch changes key
variable names so that the variables are not replaced and the
project can be built.

Signed-off-by: Bryan Evenson <bevenson@melinkcorp.com>
---
 Makefile                  |   23 +++++++++++++----------
 demo/Makefile             |   26 +++++++++++++-------------
 demo/dcc/Makefile         |    4 ++--
 demo/epics/Makefile       |    4 ++--
 demo/gateway/Makefile     |    8 ++++----
 demo/iamrouter/Makefile   |    4 ++--
 demo/initrouter/Makefile  |    4 ++--
 demo/mstpcap/Makefile     |    4 ++--
 demo/mstpcrc/Makefile     |    4 ++--
 demo/ptransfer/Makefile   |    4 ++--
 demo/readbdt/Makefile     |    4 ++--
 demo/readfile/Makefile    |    4 ++--
 demo/readprop/Makefile    |    4 ++--
 demo/readpropm/Makefile   |    4 ++--
 demo/reinit/Makefile      |    4 ++--
 demo/scov/Makefile        |    4 ++--
 demo/server/Makefile      |    4 ++--
 demo/timesync/Makefile    |    4 ++--
 demo/ucov/Makefile        |    4 ++--
 demo/uptransfer/Makefile  |    4 ++--
 demo/whohas/Makefile      |    4 ++--
 demo/whois/Makefile       |    4 ++--
 demo/whoisrouter/Makefile |    4 ++--
 demo/writefile/Makefile   |    4 ++--
 demo/writeprop/Makefile   |    4 ++--
 lib/Makefile              |   24 ++++++++++++------------
 26 files changed, 86 insertions(+), 83 deletions(-)

diff --git a/Makefile b/Makefile
index ac2dc6a..a10fba6 100644
--- a/Makefile
+++ b/Makefile
@@ -9,6 +9,9 @@
 #
 # Assumes rm and cp are available
 
+# Set the base directory.
+BACNET_TOP:=${CURDIR}
+
 # configuration
 # If BACNET_DEFINES has not already been set, configure to your needs here
 MY_BACNET_DEFINES = -DPRINT_ENABLED=1
@@ -28,27 +31,27 @@ BBMD_DEFINE ?=-DBBMD_ENABLED=1
 #BBMD_DEFINE ?= -DBBMD_CLIENT_ENABLED
 
 # Passing parameters via command line
-MAKE_DEFINE ?=
+BACNET_MAKE_DEFINE ?=
 
 # Define WEAK_FUNC for [...somebody help here; I can't find any uses of it]
-DEFINES = $(BACNET_DEFINES) $(BACDL_DEFINE) $(BBMD_DEFINE) -DWEAK_FUNC=
-DEFINES += $(MAKE_DEFINE)
+BACNET_DEFINES += $(BACDL_DEFINE) $(BBMD_DEFINE) -DWEAK_FUNC=
+BACNET_DEFINES += $(BACNET_MAKE_DEFINE)
 
 # BACnet Ports Directory
 BACNET_PORT ?= linux
 
 # Default compiler settings
-OPTIMIZATION = -Os
-DEBUGGING =
-WARNINGS = -Wall -Wmissing-prototypes
+BACNET_OPTIMIZATION = -Os
+BACNET_DEBUGGING =
+BACNET_WARNINGS = -Wall -Wmissing-prototypes
 ifeq (${BUILD},debug)
-OPTIMIZATION = -O0
-DEBUGGING = -g -DDEBUG_ENABLED=1
+BACNET_OPTIMIZATION = -O0
+BACNET_DEBUGGING = -g -DDEBUG_ENABLED=1
 ifeq (${BACDL_DEFINE},-DBACDL_BIP=1)
-DEFINES += -DBIP_DEBUG
+BACNET_DEFINES += -DBIP_DEBUG
 endif
 endif
-CFLAGS  = $(WARNINGS) $(DEBUGGING) $(OPTIMIZATION) $(STANDARDS) $(INCLUDES) $(DEFINES)
+BACNET_CFLAGS  = $(BACNET_WARNINGS) $(BACNET_DEBUGGING) $(BACNET_OPTIMIZATION) $(BACNET_STANDARDS) $(BACNET_INCLUDES) $(BACNET_DEFINES)
 
 # Export the variables defined here to all subprocesses
 # (see http://www.gnu.org/software/automake/manual/make/Special-Targets.html)
diff --git a/demo/Makefile b/demo/Makefile
index 2f9550a..cedfe54 100644
--- a/demo/Makefile
+++ b/demo/Makefile
@@ -9,18 +9,18 @@
 
 # Directories
 BACNET_PORT ?= linux
-BACNET_PORT_DIR = ../../ports/${BACNET_PORT}
-BACNET_INCLUDE = ../../include
-BACNET_OBJECT = ../../demo/object
-BACNET_HANDLER = ../../demo/handler
+BACNET_PORT_DIR := ${BACNET_TOP}/ports/${BACNET_PORT}
+BACNET_INCLUDE := ${BACNET_TOP}/include
+BACNET_OBJECT := ${BACNET_TOP}/demo/object
+BACNET_HANDLER := ${BACNET_TOP}/demo/handler
 # BACnet Library
-BACNET_LIB_DIR = ../../lib
+BACNET_LIB_DIR := ${BACNET_TOP}/lib
 BACNET_LIB_NAME = bacnet
 BACNET_LIB_TARGET = $(BACNET_LIB_DIR)/lib$(BACNET_LIB_NAME).a
 # Compiler Setup
-INCLUDE1 = -I$(BACNET_PORT_DIR) -I$(BACNET_OBJECT) -I$(BACNET_HANDLER)
-INCLUDE2 = -I$(BACNET_INCLUDE)
-INCLUDES = $(INCLUDE1) $(INCLUDE2)
+BACNET_INCLUDE1 = -I$(BACNET_PORT_DIR) -I$(BACNET_OBJECT) -I$(BACNET_HANDLER)
+BACNET_INCLUDE2 = -I$(BACNET_INCLUDE)
+BACNET_INCLUDES = $(BACNET_INCLUDE1) $(BACNET_INCLUDE2)
 BACNET_LIB=-L$(BACNET_LIB_DIR),-l$(BACNET_LIB_NAME)
 ifeq (${BACNET_PORT},linux)
 PFLAGS = -pthread
@@ -37,14 +37,14 @@ TARGET_EXT = .exe
 SYSTEM_LIB=-lws2_32,-lgcc,-lm,-liphlpapi,-lwinmm
 endif
 #build for release (default) or debug
-DEBUGGING =
-OPTIMIZATION = -Os
+BACNET_DEBUGGING =
+BACNET_OPTIMIZATION = -Os
 ifeq (${BUILD},debug)
-OPTIMIZATION = -O0
-DEBUGGING = -g
+BACNET_OPTIMIZATION = -O0
+BACNET_DEBUGGING = -g
 endif
 # put all the flags together
-CFLAGS := -Wall $(DEBUGGING) $(OPTIMIZATION) $(INCLUDES) $(DEFINES)
+BACNET_CFLAGS := -Wall $(BACNET_DEBUGGING) $(BACNET_OPTIMIZATION) $(BACNET_INCLUDES) $(BACNET_DEFINES)
 LFLAGS := -Wl,$(BACNET_LIB),$(SYSTEM_LIB)
 
 .EXPORT_ALL_VARIABLES:
diff --git a/demo/dcc/Makefile b/demo/dcc/Makefile
index 1ddc898..249ceb0 100644
--- a/demo/dcc/Makefile
+++ b/demo/dcc/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/epics/Makefile b/demo/epics/Makefile
index 01ddb04..3ab6386 100644
--- a/demo/epics/Makefile
+++ b/demo/epics/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/gateway/Makefile b/demo/gateway/Makefile
index a1fe53d..053fff0 100644
--- a/demo/gateway/Makefile
+++ b/demo/gateway/Makefile
@@ -34,9 +34,9 @@ SRCS = main.c \
 
 OBJS = ${SRCS:.c=.o}
 
-DEFINES += -DBAC_ROUTING
+BACNET_DEFINES += -DBAC_ROUTING
 
-CFLAGS  = $(WARNINGS) $(DEBUGGING) $(OPTIMIZATION) $(STANDARDS) $(INCLUDES) $(DEFINES)
+BACNET_CFLAGS  = $(BACNET_WARNINGS) $(BACNET_DEBUGGING) $(BACNET_OPTIMIZATION) $(BACNET_STANDARDS) $(BACNET_INCLUDES) $(BACNET_DEFINES)
 
 all: ${BACNET_LIB_TARGET} Makefile ${TARGET_BIN}
 
@@ -51,11 +51,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/iamrouter/Makefile b/demo/iamrouter/Makefile
index 8c33208..98aa6cb 100644
--- a/demo/iamrouter/Makefile
+++ b/demo/iamrouter/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/initrouter/Makefile b/demo/initrouter/Makefile
index ae7352e..8d08b91 100644
--- a/demo/initrouter/Makefile
+++ b/demo/initrouter/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/mstpcap/Makefile b/demo/mstpcap/Makefile
index 4990493..cae8150 100644
--- a/demo/mstpcap/Makefile
+++ b/demo/mstpcap/Makefile
@@ -36,11 +36,11 @@ ${TARGET_BIN}: ${OBJS} Makefile
 	cp $@ ../../bin
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} $(TARGET).map
diff --git a/demo/mstpcrc/Makefile b/demo/mstpcrc/Makefile
index 05e0011..0726d4b 100644
--- a/demo/mstpcrc/Makefile
+++ b/demo/mstpcrc/Makefile
@@ -41,11 +41,11 @@ ${TARGET_BIN}: ${OBJS} Makefile
 	cp $@ ../../bin
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} $(TARGET).map
diff --git a/demo/ptransfer/Makefile b/demo/ptransfer/Makefile
index b1b9867..4f4c400 100644
--- a/demo/ptransfer/Makefile
+++ b/demo/ptransfer/Makefile
@@ -32,11 +32,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -rf core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET}
diff --git a/demo/readbdt/Makefile b/demo/readbdt/Makefile
index bed0d5c..5a42846 100644
--- a/demo/readbdt/Makefile
+++ b/demo/readbdt/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/readfile/Makefile b/demo/readfile/Makefile
index 235f65c..16380fa 100644
--- a/demo/readfile/Makefile
+++ b/demo/readfile/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/readprop/Makefile b/demo/readprop/Makefile
index 836ee01..d98732c 100644
--- a/demo/readprop/Makefile
+++ b/demo/readprop/Makefile
@@ -32,11 +32,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -rf core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET}
diff --git a/demo/readpropm/Makefile b/demo/readpropm/Makefile
index 271aa73..3f08fae 100644
--- a/demo/readpropm/Makefile
+++ b/demo/readpropm/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/reinit/Makefile b/demo/reinit/Makefile
index 42984e9..4d06c80 100644
--- a/demo/reinit/Makefile
+++ b/demo/reinit/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/scov/Makefile b/demo/scov/Makefile
index 763de59..538c5c1 100644
--- a/demo/scov/Makefile
+++ b/demo/scov/Makefile
@@ -32,11 +32,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -rf core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET}
diff --git a/demo/server/Makefile b/demo/server/Makefile
index 2e26c75..5a0fcba 100644
--- a/demo/server/Makefile
+++ b/demo/server/Makefile
@@ -46,11 +46,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/timesync/Makefile b/demo/timesync/Makefile
index 28ce904..ee526ad 100644
--- a/demo/timesync/Makefile
+++ b/demo/timesync/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/ucov/Makefile b/demo/ucov/Makefile
index 703cd74..d80ff9f 100644
--- a/demo/ucov/Makefile
+++ b/demo/ucov/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/uptransfer/Makefile b/demo/uptransfer/Makefile
index 8281c24..5964881 100644
--- a/demo/uptransfer/Makefile
+++ b/demo/uptransfer/Makefile
@@ -32,11 +32,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -rf core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET}
diff --git a/demo/whohas/Makefile b/demo/whohas/Makefile
index 57387a3..09b53d5 100644
--- a/demo/whohas/Makefile
+++ b/demo/whohas/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM  ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/whois/Makefile b/demo/whois/Makefile
index 2c11748..9fb7cdb 100644
--- a/demo/whois/Makefile
+++ b/demo/whois/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/whoisrouter/Makefile b/demo/whoisrouter/Makefile
index 4940d32..bb819fb 100644
--- a/demo/whoisrouter/Makefile
+++ b/demo/whoisrouter/Makefile
@@ -26,11 +26,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/writefile/Makefile b/demo/writefile/Makefile
index f26b1fe..87e37a4 100644
--- a/demo/writefile/Makefile
+++ b/demo/writefile/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/demo/writeprop/Makefile b/demo/writeprop/Makefile
index 5a9587f..e2f894f 100644
--- a/demo/writeprop/Makefile
+++ b/demo/writeprop/Makefile
@@ -27,11 +27,11 @@ ${BACNET_LIB_TARGET}:
 	( cd ${BACNET_LIB_DIR} ; $(MAKE) clean ; $(MAKE) )
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -f core ${TARGET_BIN} ${OBJS} ${BACNET_LIB_TARGET} $(TARGET).map
diff --git a/lib/Makefile b/lib/Makefile
index b4ad476..de9f97d 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -9,16 +9,16 @@
 #
 # Assumes rm and cp are available
 
-BACNET_PORT_DIR = ../ports/${BACNET_PORT}
-BACNET_OBJECT = ../demo/object
-BACNET_HANDLER = ../demo/handler
-BACNET_CORE = ../src
-BACNET_INCLUDE = ../include
+BACNET_PORT_DIR := ${BACNET_TOP}/ports/${BACNET_PORT}
+BACNET_OBJECT := ${BACNET_TOP}/demo/object
+BACNET_HANDLER := ${BACNET_TOP}/demo/handler
+BACNET_CORE := ${BACNET_TOP}/src
+BACNET_INCLUDE := ${BACNET_TOP}/include
 # compiler configuration
 #STANDARDS = -std=c99
-INCLUDE1 = -I$(BACNET_PORT_DIR) -I$(BACNET_OBJECT) -I$(BACNET_HANDLER)
-INCLUDE2 = -I$(BACNET_INCLUDE)
-INCLUDES = $(INCLUDE1) $(INCLUDE2)
+BACNET_INCLUDE1 = -I$(BACNET_PORT_DIR) -I$(BACNET_OBJECT) -I$(BACNET_HANDLER)
+BACNET_INCLUDE2 = -I$(BACNET_INCLUDE)
+BACNET_INCLUDES = $(BACNET_INCLUDE1) $(BACNET_INCLUDE2)
 
 # target
 TARGET = bacnet
@@ -142,7 +142,7 @@ PORT_MSTP_SRC = \
 PORT_ETHERNET_SRC = \
 	$(BACNET_PORT_DIR)/ethernet.c
 
-PORT_BIP_SRC = \
+PORT_BIP_SRC := \
 	$(BACNET_PORT_DIR)/bip-init.c \
 	$(BACNET_CORE)/bvlc.c \
 	$(BACNET_CORE)/bip.c
@@ -182,7 +182,7 @@ SRCS = ${CORE_SRC} ${PORT_SRC} ${HANDLER_SRC}
 OBJS = ${SRCS:.c=.o}
 
 # use local includes, but other values from calling Makefile
-CFLAGS  = $(WARNINGS) $(DEBUGGING) $(OPTIMIZATION) $(STANDARDS) $(INCLUDES) $(DEFINES)
+BACNET_CFLAGS  = $(BACNET_WARNINGS) $(BACNET_DEBUGGING) $(BACNET_OPTIMIZATION) $(BACNET_STANDARDS) $(BACNET_INCLUDES) $(BACNET_DEFINES)
 
 all: $(LIBRARY)
 
@@ -192,11 +192,11 @@ $(LIBRARY): $(OBJS) Makefile
 	$(AR) rcs $@ $(OBJS)
 
 .c.o:
-	${CC} -c ${CFLAGS} $*.c -o $@
+	${CC} -c ${BACNET_CFLAGS} ${CFLAGS} $*.c -o $@
 
 depend:
 	rm -f .depend
-	${CC} -MM ${CFLAGS} *.c >> .depend
+	${CC} -MM ${BACNET_CFLAGS} ${CFLAGS} *.c >> .depend
 
 clean:
 	rm -rf core $(OBJS) $(LIBRARY)
-- 
1.7.9.5

